beacon_node/eth1_chain/src/eth1.rs

```diff
@@ -0,0 +1,118 @@
+use crate::cache::*;
+use crate::deposits::*;
+use crate::types::Eth1DataFetcher;
+use slog::{info, o, warn};
+use std::cmp::Ordering;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use tokio::runtime::TaskExecutor;
+use tokio::timer::Interval;
+use web3::futures::{Future, Stream};
+
+use types::*;
+
+const ETH1_FOLLOW_DISTANCE: u64 = 1024; // Need to move this to eth2_config.toml
+
+#[derive(Clone, Debug)]
+pub struct Eth1<F: Eth1DataFetcher> {
+    /// Cache for storing block_number to Eth1Data from the deposit contract.
+    pub eth1_data_cache: Eth1DataCache<F>,
+    /// Cache for storing deposit_index to Deposits received from deposit contract.
+    pub deposit_cache: DepositCache<F>,
+    /// Eth1 data fetcher
+    fetcher: Arc<F>,
+}
+
+impl<F: Eth1DataFetcher> Eth1<F> {
+    pub fn new(fetcher: F) -> Self {
+        let fetcher_arc = Arc::new(fetcher);
+        Eth1 {
+            eth1_data_cache: Eth1DataCache::new(fetcher_arc.clone()),
+            deposit_cache: DepositCache::new(fetcher_arc.clone()),
+            fetcher: fetcher_arc,
+        }
+    }
+
+    /// Get Eth1Votes with highest votes in given voting period.
+    /// From https://github.com/ethereum/eth2.0-specs/blob/v0.8.1/specs/validator/0_beacon-chain-validator.md#eth1-data
+    pub fn get_eth1_votes<T: EthSpec>(
+        &mut self,
+        state: BeaconState<T>,
+        previous_eth1_distance: u64,
+    ) -> Eth1Data {
+        let new_eth1_data = self
+            .eth1_data_cache
+            .get_eth1_data_in_range(ETH1_FOLLOW_DISTANCE, 2 * ETH1_FOLLOW_DISTANCE);
+        let all_eth1_data = self
+            .eth1_data_cache
+            .get_eth1_data_in_range(ETH1_FOLLOW_DISTANCE, previous_eth1_distance);
+        let mut valid_votes: Vec<Eth1Data> = vec![];
+        for (slot, vote) in state.eth1_data_votes.iter().enumerate() {
+            let period_tail: bool = (slot as u64 % T::SlotsPerEth1VotingPeriod::to_u64())
+                >= ((T::SlotsPerEth1VotingPeriod::to_u64() as f64).sqrt() as u64 + 1);
+            if new_eth1_data.contains(vote) || (period_tail && all_eth1_data.contains(vote)) {
+                valid_votes.push(vote.clone());
+            }
+        }
+        valid_votes
```

Comment:
@michaelsproul do we already have this logic in `state_processing`?

Reply:
I don't believe we do, at the moment we just stub out the eth1 data with dummy values